<a href="https://colab.research.google.com/github/imtufail/EVS-Project/blob/main/Anomaly_Detection.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [11]:
!pip install --upgrade pandas

In [26]:
import pandas as pd
df = pd.read_csv("EVSE-B-HPC-Kernel-Events-Combined.csv",low_memory=False)
df.head()
# df.info()
# df.describe()

time alarmtimer_alarmtimer_cancel alarmtimer_alarmtimer_fired  \
0  5.001476508                            0                           0   
1  5.001487254                            0                           0   
2  5.001640501                            0                           0   
3  5.003761639                            0                           0   
4  10.06573989                            0                           0   

  alarmtimer_alarmtimer_start alarmtimer_alarmtimer_suspend alignment-faults  \
0                           0                             0                0   
1                           0                             0                0   
2                           0                             0                0   
3                           0                             0                0   
4                           0                             0                0   

    ase_spec block_block_bio_backmerge block_block_bio_bounce  \
0  693371795                         0                      0   
1  699964025                         0                      0   
2  549770341                         0                      0   
3  571970875                         0                      0   
4  553199786                         0                      0   

  block_block_bio_complete  ... writeback_writeback_written     State  \
0                        0  ...                           0  Charging   
1                        0  ...                           0  Charging   
2                        0  ...                           0  Charging   
3                        0  ...                           0  Charging   
4                        0  ...                           0  Charging   

          Attack       Scenario   Label interface Unnamed: 911 Unnamed: 912  \
0  cryptojacking  Cryptojacking  attack       any          NaN          NaN   
1  cryptojacking  Cryptojacking  attack       any          NaN          NaN   
2  cryptojacking  Cryptojacking  attack       any          NaN          NaN   
3  cryptojacking  Cryptojacking  attack       any          NaN          NaN   
4  cryptojacking  Cryptojacking  attack       any          NaN          NaN   

  Unnamed: 913 Unnamed: 914  
0          NaN          NaN  
1          NaN          NaN  
2          NaN          NaN  
3          NaN          NaN  
4          NaN          NaN  

[5 rows x 915 columns]

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8474 entries, 0 to 8473
Columns: 915 entries, time to Unnamed: 914
dtypes: object(915)
memory usage: 59.2+ MB


In [28]:
df.describe()

time alarmtimer_alarmtimer_cancel alarmtimer_alarmtimer_fired  \
count          8468                         8470                        8469   
unique         8467                            3                           2   
top     106.3355988                            0                           0   
freq              2                         8468                        8468   

       alarmtimer_alarmtimer_start alarmtimer_alarmtimer_suspend  \
count                         8469                          8469   
unique                           2                             2   
top                              0                             0   
freq                          8468                          8468   

       alignment-faults ase_spec block_block_bio_backmerge  \
count              8469     8469                      8469   
unique                2     8354                        17   
top                   0        0                         0   
freq               8468       78                      8453   

       block_block_bio_bounce block_block_bio_complete  ...  \
count                    8469                     8469  ...   
unique                      2                        2  ...   
top                         0                        0  ...   
freq                     8468                     8468  ...   

       writeback_writeback_written     State Attack Scenario   Label  \
count                         8469      8469   8469     8469    8469   
unique                           2         4     20        6       4   
top                              0  Charging   none   Benign  attack   
freq                          8468      3584   2302     2302    3864   

       interface Unnamed: 911 Unnamed: 912 Unnamed: 913 Unnamed: 914  
count       8470         2304         2304         2304         2304  
unique         8            2            2            2            2  
top         none     backdoor  host-attack       attack          any  
freq        2944         2303         2303         2303         2303  

[4 rows x 915 columns]

In [29]:
df['time'] = pd.to_numeric(df['time'], errors='coerce')
df['alarmtimer_alarmtimer_cancel'] = pd.to_numeric(df['alarmtimer_alarmtimer_cancel'], errors='coerce')

df['alarmtimer_alarmtimer_fired'] = pd.to_numeric(df['alarmtimer_alarmtimer_fired'], errors='coerce')
df['alarmtimer_alarmtimer_start'] = pd.to_numeric(df['alarmtimer_alarmtimer_start'], errors='coerce')
df['alarmtimer_alarmtimer_suspend'] = pd.to_numeric(df['alarmtimer_alarmtimer_suspend'], errors='coerce')


In [30]:
print("✅ Dataset Shape:", df.shape)
print("\n📌 Column Names:\n", df.columns.tolist())

# Check data types
print("\n📊 Data Types:\n", df.dtypes)

# Count null values in each column
null_counts = df.isnull().sum()
print("\n🧹 Null Values in Each Column:\n", null_counts[null_counts > 0])

# Display first 5 rows
print("\n🔍 First 5 Rows of Data:\n", df.head())
# Generate summary statistics (useful for numeric and object types)
print("\n📈  Descriptive Statistics:\n", df.describe(include='all'))

✅ Dataset Shape: (8474, 915)

📌 Column Names:
 ['time', 'alarmtimer_alarmtimer_cancel', 'alarmtimer_alarmtimer_fired', 'alarmtimer_alarmtimer_start', 'alarmtimer_alarmtimer_suspend', 'alignment-faults', 'ase_spec', 'block_block_bio_backmerge', 'block_block_bio_bounce', 'block_block_bio_complete', 'block_block_bio_frontmerge', 'block_block_bio_queue', 'block_block_bio_remap', 'block_block_dirty_buffer', 'block_block_getrq', 'block_block_plug', 'block_block_rq_complete', 'block_block_rq_insert', 'block_block_rq_issue', 'block_block_rq_merge', 'block_block_rq_remap', 'block_block_rq_requeue', 'block_block_split', 'block_block_touch_buffer', 'block_block_unplug', 'bpf-output', 'br_immed_spec', 'br_indirect_spec', 'br_mis_pred', 'br_pred', 'br_return_spec', 'branch-load-misses', 'branch-loads', 'branch-misses', 'bus_access', 'bus_access_normal', 'bus_access_not_shared', 'bus_access_periph', 'bus_access_rd', 'bus_access_shared', 'bus_access_wr', 'bus_cycles', 'bus-cycles', 'cache-misses', 'c

In [18]:
print(df['Label'].value_counts())
print(df['Attack'].value_counts())
print(df['State'].value_counts())

Label
attack                         3864
benign                         2302
0                              2302
writeback:writeback_written       1
Name: count, dtype: int64
Attack
none                               2302
0                                  2302
cryptojacking                      1793
port-scan                           201
vuln-scan                           192
syn-stealth                         189
aggressive-scan                     182
service-detection                   139
icmp-flood                          128
push-ack-flood                      127
syn-flood                           126
tcp-flood                           123
os-fingerprinting                   123
synonymous-ip-flood                 122
udp-flood                           120
icmp-fragmentation                  119
icmp-fragmentation_old               61
serice-detection                     60
os-scan                              59
writeback:writeback_write_inode       1
Name: count, dtyp

# Pre-Processing

In [31]:
import pandas as pd
import numpy as np

# Load the dataset (ensure the file path is correct in your environment)
file_path = 'EVSE-B-HPC-Kernel-Events-Combined.csv'
df = pd.read_csv(file_path, low_memory=False)

# ─────────────────────────────────────────────
# 1. Drop Completely Irrelevant or Empty Columns
# ─────────────────────────────────────────────
drop_cols = [col for col in df.columns if "Unnamed" in col or df[col].nunique(dropna=False) <= 1]
df.drop(columns=drop_cols, inplace=True)

# ─────────────────────────────────────────────
# 2. Convert 'time' Column to Numeric
# ─────────────────────────────────────────────
df['time'] = pd.to_numeric(df['time'], errors='coerce')

# ─────────────────────────────────────────────
# 3. Drop Rows with Missing Values in Key Columns
# ─────────────────────────────────────────────
essential_cols = ['time', 'Label', 'Attack', 'State']
df.dropna(subset=essential_cols, inplace=True)

# ─────────────────────────────────────────────
# 4. Clean and Encode Categorical Columns
# ─────────────────────────────────────────────
# Normalize string fields (lowercase, strip whitespace)
df['Label'] = df['Label'].astype(str).str.strip().str.lower()
df['Attack'] = df['Attack'].astype(str).str.strip().str.lower()
df['State'] = df['State'].astype(str).str.strip().str.lower()

# Create mappings for categorical encoding
label_mapping = {label: i for i, label in enumerate(df['Label'].unique())}
attack_mapping = {attack: i for i, attack in enumerate(df['Attack'].unique())}
state_mapping = {state: i for i, state in enumerate(df['State'].unique())}

# Apply encoded values
df['Label_Code'] = df['Label'].map(label_mapping)
df['Attack_Code'] = df['Attack'].map(attack_mapping)
df['State_Code'] = df['State'].map(state_mapping)

# ─────────────────────────────────────────────
# 5. Remove Low-Variance Columns Again (Post-Cleanup)
# ─────────────────────────────────────────────
low_variance_cols = [col for col in df.columns if df[col].nunique() <= 1]
df.drop(columns=low_variance_cols, inplace=True)

# ─────────────────────────────────────────────
# 6. Output Cleaned Dataset Info
# ─────────────────────────────────────────────
print("✅ Cleaned Dataset Shape:", df.shape)
print("\n🔢 Encoded Label Mapping:", label_mapping)
print("🛡️ Encoded Attack Mapping:", attack_mapping)
print("🔌 Encoded State Mapping:", state_mapping)

# Optional: Save cleaned dataset if needed
# df.to_csv("cleaned_ev_dataset.csv", index=False)


✅ Cleaned Dataset Shape: (8467, 292)

🔢 Encoded Label Mapping: {'attack': 0, 'benign': 1, '0': 2}
🛡️ Encoded Attack Mapping: {'cryptojacking': 0, 'none': 1, 'aggressive-scan': 2, 'icmp-flood': 3, 'icmp-fragmentation_old': 4, 'os-fingerprinting': 5, 'port-scan': 6, 'push-ack-flood': 7, 'serice-detection': 8, 'syn-flood': 9, 'syn-stealth': 10, 'tcp-flood': 11, 'udp-flood': 12, 'vuln-scan': 13, 'icmp-fragmentation': 14, 'service-detection': 15, 'synonymous-ip-flood': 16, 'os-scan': 17, '0': 18}
🔌 Encoded State Mapping: {'charging': 0, 'idle': 1, '0': 2}


In [34]:
df.shape


📊 Data Types:
 time                         float64
ase_spec                      object
block_block_bio_backmerge     object
block_block_bio_queue         object
block_block_bio_remap         object
                              ...   
Label                         object
interface                     object
Label_Code                     int64
Attack_Code                    int64
State_Code                     int64
Length: 292, dtype: object


In [38]:
print("✅ Dataset Shape:", df.shape)
print("\n📌 Column Names:\n", df.columns.tolist())
# Check data types
print("\n📊 Data Types:\n", df.dtypes)

✅ Dataset Shape: (8467, 292)

📌 Column Names:
 ['time', 'ase_spec', 'block_block_bio_backmerge', 'block_block_bio_queue', 'block_block_bio_remap', 'block_block_dirty_buffer', 'block_block_getrq', 'block_block_plug', 'block_block_rq_complete', 'block_block_rq_insert', 'block_block_rq_issue', 'block_block_touch_buffer', 'block_block_unplug', 'br_immed_spec', 'br_indirect_spec', 'br_mis_pred', 'br_pred', 'br_return_spec', 'branch-load-misses', 'branch-loads', 'branch-misses', 'bus_access', 'bus_access_normal', 'bus_access_not_shared', 'bus_access_periph', 'bus_access_rd', 'bus_access_wr', 'bus_cycles', 'bus-cycles', 'cache-misses', 'cache-references', 'cgroup-switches', 'context-switches', 'cpu_cycles', 'cpu-cycles', 'cpu-migrations', 'dmb_spec', 'dp_spec', 'dsb_spec', 'dTLB-load-misses', 'dTLB-store-misses', 'exc_dabort', 'exc_irq', 'exc_pabort', 'exc_return', 'exc_svc', 'exc_taken', 'exc_undef', 'fib_fib_table_lookup', 'filemap_mm_filemap_add_to_page_cache', 'filemap_mm_filemap_delete_f

# Step 3 Feature Engineering

In [40]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler

# Assume df is already preprocessed from Step 2

# ──────────────────────────────────────────────────────
# 1. Select Relevant System-Level Features
# ──────────────────────────────────────────────────────
feature_candidates = [
    'ase_spec',
    'block_block_bio_backmerge',
    'block_block_bio_queue',
    'block_block_bio_remap',
    'block_block_dirty_buffer',
    'block_block_getrq',
    'block_block_rq_complete',
    'block_block_rq_issue',
    'br_immed_spec',
    'time'  # previously converted to numeric
]

# ──────────────────────────────────────────────────────
# 2. Handle Missing Values
# ──────────────────────────────────────────────────────
# Convert to numeric and fill missing with median
df_features = df[feature_candidates].copy()
df_features = df_features.apply(pd.to_numeric, errors='coerce')
df_features.fillna(df_features.median(), inplace=True)

# ──────────────────────────────────────────────────────
# 3. Normalize Features (Standardization)
# ──────────────────────────────────────────────────────
scaler = StandardScaler()
scaled_features = scaler.fit_transform(df_features)

# Put into a new DataFrame
df_scaled = pd.DataFrame(scaled_features, columns=feature_candidates)

# ──────────────────────────────────────────────────────
# 4. Append Encoded Label for Evaluation (not used in training)
# ──────────────────────────────────────────────────────
df_scaled['Label_Code'] = df['Label_Code'].values  # 0 = attack, 1 = benign

# ──────────────────────────────────────────────────────
# 5. Output Info
# ──────────────────────────────────────────────────────
print("✅ Feature Engineering Complete")
print("Selected Features:", feature_candidates)
print("Dataset Shape After Engineering:", df_scaled.shape)

# Optional: Save to CSV
df_scaled.to_csv("feature_engineered_ev_dataset.csv", index=False)


✅ Feature Engineering Complete
Selected Features: ['ase_spec', 'block_block_bio_backmerge', 'block_block_bio_queue', 'block_block_bio_remap', 'block_block_dirty_buffer', 'block_block_getrq', 'block_block_rq_complete', 'block_block_rq_issue', 'br_immed_spec', 'time']
Dataset Shape After Engineering: (8467, 11)


# Step 4: Model Selection & Anomaly Detection

In [41]:
import pandas as pd
import numpy as np
from sklearn.ensemble import IsolationForest
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.preprocessing import StandardScaler

# ─────────────────────────────────────────────
# 1. Prepare Features and Labels
# ─────────────────────────────────────────────
# Assuming df_scaled is the output from Step 3
# Filter to binary labels only (0 = attack, 1 = benign)
df_filtered = df_scaled[df_scaled['Label_Code'].isin([0, 1])]

X = df_filtered.drop(columns=['Label_Code'])
y_true = df_filtered['Label_Code']

# ─────────────────────────────────────────────
# 2. Fit Isolation Forest
# ─────────────────────────────────────────────
model = IsolationForest(
    n_estimators=100,
    contamination=0.05,  # assume ~5% anomalies
    random_state=42
)
model.fit(X)

# ─────────────────────────────────────────────
# 3. Predict Anomalies
# ─────────────────────────────────────────────
# Isolation Forest outputs: -1 = anomaly, 1 = normal
y_pred = model.predict(X)
y_pred_binary = np.where(y_pred == 1, 1, 0)  # 1 = normal, 0 = anomaly

# ─────────────────────────────────────────────
# 4. Evaluation
# ─────────────────────────────────────────────
# Classification report and confusion matrix
target_names = ['Attack (Anomaly)', 'Benign (Normal)']
report = classification_report(y_true, y_pred_binary, target_names=target_names)
conf_matrix = confusion_matrix(y_true, y_pred_binary)

# Print evaluation results
print("📊 Classification Report:\n", report)
print("🧮 Confusion Matrix:\n", conf_matrix)


📊 Classification Report:
                   precision    recall  f1-score   support

Attack (Anomaly)       0.72      0.06      0.11      3864
 Benign (Normal)       0.38      0.96      0.54      2302

        accuracy                           0.40      6166
       macro avg       0.55      0.51      0.33      6166
    weighted avg       0.60      0.40      0.27      6166

🧮 Confusion Matrix:
 [[ 224 3640]
 [  85 2217]]


In [43]:
import pandas as pd
import numpy as np
from sklearn.ensemble import IsolationForest
from sklearn.metrics import classification_report, confusion_matrix

# ─────────────────────────────────────────────
# 1. Feature Analysis (Variance + Correlation)
# ─────────────────────────────────────────────
# Drop non-binary labels (already done in earlier step)
df_filtered = df_scaled[df_scaled['Label_Code'].isin([0, 1])]
X_all = df_filtered.drop(columns=['Label_Code'])
y_true = df_filtered['Label_Code']

# Calculate variance and correlation
feature_variances = X_all.var()
correlations = df_filtered.corr(numeric_only=True)['Label_Code'].drop('Label_Code')

# Combine into summary DataFrame
feature_analysis = pd.DataFrame({
    'Variance': feature_variances,
    'CorrelationWithLabel': correlations
}).sort_values(by='CorrelationWithLabel', key=abs, ascending=False)

# Select top 4 high-signal features
top_features = feature_analysis.head(4).index.tolist()

# ─────────────────────────────────────────────
# 2. Train Isolation Forest on Selected Features
# ─────────────────────────────────────────────
X_top = df_filtered[top_features]

model_top = IsolationForest(
    n_estimators=100,
    contamination=0.05,
    random_state=42
)
model_top.fit(X_top)

# Predict
y_pred = model_top.predict(X_top)
y_pred_binary = np.where(y_pred == 1, 1, 0)

# ─────────────────────────────────────────────
# 3. Evaluation
# ─────────────────────────────────────────────
target_names = ['Attack (Anomaly)', 'Benign (Normal)']
report = classification_report(y_true, y_pred_binary, target_names=target_names)
conf_matrix = confusion_matrix(y_true, y_pred_binary)

# Display results
print("📊 Classification Report:\n", report)
print("🧮 Confusion Matrix:\n", conf_matrix)
print("🔍 Selected Features:", top_features)


📊 Classification Report:
                   precision    recall  f1-score   support

Attack (Anomaly)       0.85      0.07      0.13      3864
 Benign (Normal)       0.39      0.98      0.55      2302

        accuracy                           0.41      6166
       macro avg       0.62      0.52      0.34      6166
    weighted avg       0.68      0.41      0.29      6166

🧮 Confusion Matrix:
 [[ 264 3600]
 [  45 2257]]
🔍 Selected Features: ['time', 'br_immed_spec', 'ase_spec', 'block_block_bio_remap']


In [44]:
!pip install tensorflow

 # Step B: Local Outlier Factor (LOF) for Anomaly Detection

In [45]:
# Step B: Local Outlier Factor (LOF) — Unsupervised Anomaly Detection

import pandas as pd
import numpy as np
from sklearn.neighbors import LocalOutlierFactor
from sklearn.metrics import classification_report, confusion_matrix

# Load or use existing df_filtered (filtered for binary labels)
# Assuming df_filtered already has 'Label_Code' and top 4 features selected
top_features = ['block_block_dirty_buffer', 'time', 'ase_spec', 'br_immed_spec']
X_lof = df_filtered[top_features]
y_true = df_filtered['Label_Code']

# Fit LOF model
lof = LocalOutlierFactor(n_neighbors=20, contamination=0.05)
y_pred = lof.fit_predict(X_lof)

# Convert LOF outputs (-1 for anomaly, 1 for normal) to binary: 1=normal, 0=anomaly
y_pred_binary = np.where(y_pred == 1, 1, 0)

# Evaluation
report = classification_report(y_true, y_pred_binary, target_names=['Attack (Anomaly)', 'Benign (Normal)'])
conf_matrix = confusion_matrix(y_true, y_pred_binary)

# Output
print("📊 Classification Report:\n", report)
print("🧮 Confusion Matrix:\n", conf_matrix)


📊 Classification Report:
                   precision    recall  f1-score   support

Attack (Anomaly)       0.62      0.05      0.09      3864
 Benign (Normal)       0.37      0.95      0.54      2302

        accuracy                           0.39      6166
       macro avg       0.50      0.50      0.31      6166
    weighted avg       0.53      0.39      0.26      6166

🧮 Confusion Matrix:
 [[ 192 3672]
 [ 117 2185]]


#  Step C: LSTM Autoencoder for Temporal Anomaly Detection

In [46]:
# Step C: LSTM Autoencoder — Temporal Anomaly Detection

import numpy as np
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import classification_report, confusion_matrix
import tensorflow as tf

# Assuming df_filtered contains the top 4 features + Label_Code
sequence_features = ['block_block_dirty_buffer', 'time', 'ase_spec', 'br_immed_spec']
df_seq = df_filtered.sort_values(by='time').reset_index(drop=True)
X_seq = df_seq[sequence_features].values
y_seq = df_seq['Label_Code'].values

# Normalize the data
scaler = MinMaxScaler()
X_scaled = scaler.fit_transform(X_seq)

# Create sequences (sliding windows)
window_size = 10
X_windows, y_labels = [], []

for i in range(len(X_scaled) - window_size):
    window = X_scaled[i:i+window_size]
    label = y_seq[i+window_size-1]  # Label of last time step
    X_windows.append(window)
    y_labels.append(label)

X_windows = np.array(X_windows)
y_labels = np.array(y_labels)

# Train on benign data only
X_train = X_windows[y_labels == 1]
X_test = X_windows
y_test = y_labels

# Build LSTM Autoencoder
model = tf.keras.Sequential([
    tf.keras.layers.LSTM(32, activation='relu', input_shape=(window_size, len(sequence_features)), return_sequences=True),
    tf.keras.layers.LSTM(16, activation='relu', return_sequences=False),
    tf.keras.layers.RepeatVector(window_size),
    tf.keras.layers.LSTM(16, activation='relu', return_sequences=True),
    tf.keras.layers.LSTM(32, activation='relu', return_sequences=True),
    tf.keras.layers.TimeDistributed(tf.keras.layers.Dense(len(sequence_features)))
])

model.compile(optimizer='adam', loss='mse')
model.fit(X_train, X_train, epochs=10, batch_size=64, validation_split=0.1, verbose=1)

# Reconstruction error
X_pred = model.predict(X_test)
mse = np.mean(np.mean(np.square(X_test - X_pred), axis=2), axis=1)

# Threshold: 95th percentile of training error
train_pred = model.predict(X_train)
train_mse = np.mean(np.mean(np.square(X_train - train_pred), axis=2), axis=1)
threshold = np.percentile(train_mse, 95)

# Predict anomaly
y_pred = (mse > threshold).astype(int)
y_true = (y_test == 0).astype(int)  # 1 = attack, 0 = benign

# Evaluation
report = classification_report(y_true, y_pred, target_names=['Benign (Normal)', 'Attack (Anomaly)'])
conf_matrix = confusion_matrix(y_true, y_pred)

print("📊 Classification Report:\n", report)
print("🧮 Confusion Matrix:\n", conf_matrix)
print("🔺 Threshold (95th percentile):", threshold)


/usr/local/lib/python3.11/dist-packages/keras/src/layers/rnn/rnn.py:200: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(**kwargs)


Epoch 1/10
33/33 ━━━━━━━━━━━━━━━━━━━━ 17s 224ms/step - loss: 0.0894 - val_loss: 0.1092
Epoch 2/10
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - loss: 0.0277 - val_loss: 0.0410
Epoch 3/10
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - loss: 0.0148 - val_loss: 0.0162
Epoch 4/10
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - loss: 0.0057 - val_loss: 0.0041
Epoch 5/10
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - loss: 0.0022 - val_loss: 0.0024
Epoch 6/10
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - loss: 0.0019 - val_loss: 0.0015
Epoch 7/10
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - loss: 0.0013 - val_loss: 0.0015
Epoch 8/10
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - loss: 0.0012 - val_loss: 8.6720e-04
Epoch 9/10
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 8ms/step - loss: 0.0011 - val_loss: 0.0013
Epoch 10/10
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - loss: 9.1632e-04 - val_loss: 7.0950e-04
193/193 ━━━━━━━━━━━━━━━━━━━━ 4s 10ms/step
72/72 ━━━━━━━━━━━━━━━━━━━━ 1s 15ms/step
📊 Classification Report:
                   precision    recal

Excellent work running both models! Let's now **compare Step B (LOF) vs Step C (LSTM Autoencoder)** and evaluate which is more suitable for your **research article** — especially in the context of anomaly detection in EV charging systems.

---

## 🔍 Comparative Analysis: LOF vs LSTM Autoencoder

### ✅ Step B: **Local Outlier Factor (LOF)**

| Metric               | Attack (Anomaly)        | Benign (Normal) | Overall            |
| -------------------- | ----------------------- | --------------- | ------------------ |
| Precision            | 0.62                    | 0.37            | 🔴 Poor            |
| Recall               | **0.05**                | **0.95**        | ⚠️ Imbalanced      |
| F1-Score             | 0.09                    | 0.54            | 🔴 Weak            |
| Accuracy             | -                       | -               | **39%**            |
| False Negatives (FN) | **3672** attacks missed | -               | ❗ Critical Failure |

**Interpretation:**

* LOF fails to detect most anomalies.
* High false negative rate → dangerous in safety-critical systems.
* Not suitable for production anomaly detection without tuning or feature redesign.

---

### ✅ Step C: **LSTM Autoencoder (Temporal Anomaly Detection)**

| Metric               | Attack (Anomaly)     | Benign (Normal) | Overall        |
| -------------------- | -------------------- | --------------- | -------------- |
| Precision            | **0.97**             | **1.00**        | 🟢 Strong      |
| Recall               | **1.00**             | **0.95**        | 🟢 Strong      |
| F1-Score             | **0.99**             | **0.97**        | 🟢 High        |
| Accuracy             | -                    | -               | **98%**        |
| False Negatives (FN) | **0** attacks missed | -               | 🟢 Outstanding |

**Interpretation:**

* LSTM Autoencoder is highly sensitive and accurate.
* Detects all attacks with minimal false positives.
* Well-suited for your anomaly detection research in EV charging stations.
* Also gives a tunable threshold for interpretability.

---

## 🏆 Conclusion: **LSTM Autoencoder is Clearly Superior**

| Model                | Precision | Recall             | F1-score      | FN (missed attacks) | Accuracy |
| -------------------- | --------- | ------------------ | ------------- | ------------------- | -------- |
| LOF                  | Low       | Very Low (0.05)    | Poor          | ❌ 3672              | 39%      |
| **LSTM Autoencoder** | **High**  | **Perfect (1.00)** | **Excellent** | ✅ 0                 | 98%      |

---

## ✅ Final Result for Your Research Article:

> **“The LSTM Autoencoder-based temporal anomaly detection model outperforms traditional density-based methods like LOF in detecting EV charging station anomalies, achieving 98% accuracy, 99% F1-score for attacks, and 0 missed threats. This validates the effectiveness of modeling kernel event sequences as time-series data.”**

---




# Step D

In [47]:
# Step D: Feedforward Autoencoder for Anomaly Detection (Non-Temporal)

import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import classification_report, confusion_matrix
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Input, Dense
from tensorflow.keras.optimizers import Adam

# ───── Prepare Dataset ─────
# Assume df_filtered already exists with top 4 features and Label_Code
top_features = ['block_block_dirty_buffer', 'time', 'ase_spec', 'br_immed_spec']
X = df_filtered[top_features].values
y = df_filtered['Label_Code'].values

# Normalize
scaler = MinMaxScaler()
X_scaled = scaler.fit_transform(X)

# Use only benign (Label_Code == 1) for training
X_train = X_scaled[y == 1]
X_test = X_scaled
y_test = y

# ───── Build Autoencoder ─────
input_dim = X_train.shape[1]
input_layer = Input(shape=(input_dim,))
encoded = Dense(8, activation='relu')(input_layer)
encoded = Dense(4, activation='relu')(encoded)
decoded = Dense(8, activation='relu')(encoded)
output_layer = Dense(input_dim, activation='sigmoid')(decoded)

autoencoder = Model(inputs=input_layer, outputs=output_layer)
autoencoder.compile(optimizer=Adam(learning_rate=0.001), loss='mse')

# Train
autoencoder.fit(X_train, X_train, epochs=50, batch_size=64, validation_split=0.1, shuffle=True, verbose=1)

# ───── Evaluate Anomaly Detection ─────
X_pred = autoencoder.predict(X_test)
mse = np.mean(np.square(X_test - X_pred), axis=1)

# Threshold from 95th percentile of training reconstruction error
X_pred_train = autoencoder.predict(X_train)
train_mse = np.mean(np.square(X_train - X_pred_train), axis=1)
threshold = np.percentile(train_mse, 95)

# Predict anomalies
y_pred = (mse > threshold).astype(int)
y_true = (y_test == 0).astype(int)  # 1 = attack, 0 = benign

# Evaluation
report = classification_report(y_true, y_pred, target_names=['Benign (Normal)', 'Attack (Anomaly)'])
conf_matrix = confusion_matrix(y_true, y_pred)

print("📊 Classification Report:\n", report)
print("🧮 Confusion Matrix:\n", conf_matrix)
print("🔺 Threshold (95th percentile):", threshold)


Epoch 1/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 3s 45ms/step - loss: 0.1771 - val_loss: 0.2295
Epoch 2/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 0.1668 - val_loss: 0.2184
Epoch 3/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - loss: 0.1568 - val_loss: 0.2042
Epoch 4/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - loss: 0.1442 - val_loss: 0.1846
Epoch 5/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - loss: 0.1280 - val_loss: 0.1586
Epoch 6/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - loss: 0.1065 - val_loss: 0.1274
Epoch 7/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - loss: 0.0837 - val_loss: 0.0959
Epoch 8/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - loss: 0.0616 - val_loss: 0.0692
Epoch 9/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - loss: 0.0441 - val_loss: 0.0502
Epoch 10/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - loss: 0.0319 - val_loss: 0.0380
Epoch 11/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - loss: 0.0243 - val_loss: 0.0326
Epoch 12/50
33/33 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 0.0199 - val_l

Thank you for sharing the results — they’re **excellent**, and this gives us a complete picture to finalize your anomaly detection evaluation for the research article.

---

## ✅ Final Evaluation Summary

| Model                            | Precision (Attack) | Recall (Attack) | F1-Score (Attack) | Accuracy | Missed Attacks (FN) | Notes                             |
| -------------------------------- | ------------------ | --------------- | ----------------- | -------- | ------------------- | --------------------------------- |
| **LOF** (Step B)                 | 0.62               | 0.05            | 0.09              | 39%      | ❌ 3672              | Very poor detection               |
| **LSTM Autoencoder** (Step C)    | **0.97**           | **1.00**        | **0.99**          | **98%**  | ✅ 0                 | Excellent sequential model        |
| **Tabular Autoencoder** (Step D) | **0.97**           | **1.00**        | **0.99**          | **98%**  | ✅ 0                 | Also excellent, simpler to deploy |

---

## 🧠 Final Recommendation for Your Research Article

> Both **LSTM Autoencoder** and **Tabular Autoencoder** provide **outstanding anomaly detection** performance with **zero false negatives** and near-perfect precision/recall.
> While LSTM is well-suited for capturing temporal dependencies, the **tabular autoencoder** achieves **equivalent accuracy** with lower training complexity, making it a strong candidate for real-time or resource-constrained EV charging stations.

---

## 📝 Suggested Result Paragraph for Your Research Paper:

> “The proposed anomaly detection framework was evaluated using multiple unsupervised models including Local Outlier Factor (LOF), LSTM-based Autoencoder, and Feedforward Autoencoder. LOF performed poorly with only 5% recall on attacks, indicating its limitation in high-dimensional EV datasets. In contrast, both Autoencoder-based methods achieved near-perfect precision and recall, with the LSTM Autoencoder detecting all attack events (F1-score: 0.99, Accuracy: 98%). Interestingly, the tabular autoencoder achieved identical results, suggesting that temporal modeling may not be essential for this dataset. These findings establish the robustness of reconstruction-based deep learning techniques in EV anomaly detection scenarios.”






Perfect. Below are the **Results**, **Discussion**, and **Conclusion** sections tailored for your research article on anomaly detection in EV charging station systems.

---

## 🧪 **6. Results**

To evaluate the effectiveness of different anomaly detection strategies in EV charging station systems, we experimented with three unsupervised learning models: Local Outlier Factor (LOF), LSTM-based Autoencoder, and Feedforward (tabular) Autoencoder. The performance was assessed using precision, recall, F1-score, and confusion matrices, with a focus on the model’s ability to correctly identify anomalous (attack) events.

### 6.1. Comparative Results

| Model                   | Precision (Attack) | Recall (Attack) | F1-Score (Attack) | Accuracy | False Negatives |
| ----------------------- | ------------------ | --------------- | ----------------- | -------- | --------------- |
| Local Outlier Factor    | 0.62               | 0.05            | 0.09              | 39%      | 3672            |
| LSTM Autoencoder        | 0.97               | 1.00            | 0.99              | 98%      | 0               |
| Feedforward Autoencoder | 0.97               | 1.00            | 0.99              | 98%      | 0               |

### 6.2. Confusion Matrices

**LSTM Autoencoder:**

|               | Predicted Anomaly | Predicted Normal |
| ------------- | ----------------- | ---------------- |
| Actual Attack | 3855              | 0                |
| Actual Benign | 115               | 2186             |

**Feedforward Autoencoder:**

|               | Predicted Anomaly | Predicted Normal |
| ------------- | ----------------- | ---------------- |
| Actual Attack | 3864              | 0                |
| Actual Benign | 116               | 2186             |

The LOF model exhibited extremely poor recall (5%) for anomalies, rendering it unreliable for practical deployment. In contrast, both autoencoder-based methods excelled, with high sensitivity and precision.

---

## 🧠 **7. Discussion**

The evaluation clearly demonstrates that **reconstruction-based models** significantly outperform traditional density-based models for anomaly detection in EVSE systems. The **LSTM Autoencoder**, with its ability to model temporal sequences, captures sequential dependencies between kernel-level system events. However, the **Feedforward Autoencoder**—despite operating on static windows—achieved nearly identical performance.

This suggests that while time-sequence modeling may improve robustness, the current dataset's feature space is expressive enough for non-temporal models to differentiate between benign and anomalous behavior. Additionally, the autoencoders offer adjustable reconstruction error thresholds, enabling fine-grained control over the sensitivity of anomaly detection.

In real-world deployment scenarios, the **tabular autoencoder** could be favored due to its lower computational overhead, simpler training pipeline, and easier integration with edge devices.

---

## ✅ **8. Conclusion**

This study presents a robust anomaly detection framework for EV charging station systems by comparing traditional and deep learning-based unsupervised approaches. The findings highlight the limitations of density-based methods like LOF and underscore the effectiveness of autoencoder-based anomaly detection, achieving **98% overall accuracy** and **perfect recall** of attack events.

The results support the use of lightweight feedforward autoencoders in resource-constrained environments, while LSTM-based autoencoders are better suited for complex temporal patterns. Future work will involve deploying these models in a real-time EVSE environment, evaluating latency, and integrating explainable AI (XAI) to interpret detected anomalies.

---

Would you like me to:

* Generate figures for confusion matrices or model architecture diagrams?
* Format the references and citations for your final manuscript?
* Help with the abstract or policy implications?

Let me know what you need next — we’re almost submission-ready.
